src/engine-actions/translate-swap.js

```diff
@@ -1,220 +1,74 @@
  * to the counterparty (Taker) node on this network
  *
- * @param {string} takerAddress - Payment channel network address fo the Taker node we are paying
- * @param {string} swapHash     - swap hash that will be associated with the swap
- * @param {string} amount       - Int64 string of the amount of outbound currency in integer units
- * @param {string} extendedTimeLockDelta  - Int64 string of the time lock extended to us by on the first chain in seconds
+ * @param {string} takerAddress          - Payment channel network address of
+ *                                         the Taker node we are paying
+ * @param {string} swapHash              - swap hash that will be associated
+ *                                         with the swap
+ * @param {string} amount                - Int64 string of the amount of
+ *                                         outbound currency in integer units
+ * @param {Date} maxTime                 - Absolute time representing the max
+ *                                         time for a payment to resolve
+ *                                         considering the time locks on the
+ *                                         route.
+ * @param {string} finalCltvDeltaSecs    - Int64 string of final hop time lock
+ *                                         in seconds
  * @returns {TranslateSwapOutcome}
- * @todo make time lock dynamic based on pre-agreed or advertised routes
  */
-async function translateSwap (takerAddress, swapHash, amount, extendedTimeLockDelta) {
-  this.logger.info(`Translating swap for ${swapHash} by sending to ${takerAddress}`, { amount, extendedTimeLockDelta })
+async function translateSwap (takerAddress, swapHash, amount, maxTime, finalCltvDeltaSecs = DEFAULT_MIN_FINAL_DELTA) {
+  this.logger.info(`Translating swap for ${swapHash} by sending to ${takerAddress}`,
+    { amount, maxTime })
 
   const { publicKey } = networkAddressFormatter.parse(takerAddress)
 
-  const secondsPerBlock = this.secondsPerBlock
-
-  // We specifically use Math.ceil here to ensure that the swap succeeds by providing
-  // the terminating node (Taker) additional timelock. This allows enough time so that
-  // the Taker will not reject the HTLC as having too little timelock. We add a Block buffer
-  // to account for blocks that are mined during swap execution.
-  const finalCltvDelta = Math.ceil((DEFAULT_MIN_FINAL_DELTA + BLOCK_BUFFER) / secondsPerBlock)
-
-  const queryRoutesReq = {
-    pubKey: publicKey,
+  // Convert the maximum time into a time from now so we can convert it into
+  // blocks of relative time. We round down to a whole number of seconds to
+  // be conservative.
+  const timeFromNow = Math.floor((maxTime - new Date()) / 1000)
+
+  // Translate seconds into blocks, rounding down to ensure we can stay atomic.
+  // The LND api expects this to be an int32, which in javascript is a number.
+  const cltvLimit = parseInt(Big(timeFromNow).div(this.secondsPerBlock).round(3), 10)
```

Comment:
`Big.round` returns a `Big` and `parseInt` expects a string, so this would be a type error; I think you could do
`parseInt(Big(timeFromNow).div(this.secondsPerBlock).round(0, 0).toFixed(0), 10)`

Reply:
Good catch, thanks.